server/chat.ts

```diff
@@ -1417,16 +1417,39 @@ export const Chat = new class {
 
 		// Install plug-in commands and chat filters
 
-		// info always goes first so other plugins can shadow it
-		let files = FS('server/chat-plugins/').readdirSync();
-		files = files.filter(file => file !== 'info.js');
+		// All resulting filenames will be relative to basePath
+		const getFiles = (basePath: string, path: string): string[] => {
+			const filesInThisDir = FS(`${basePath}/${path}`).readdirSync();
+			let allFiles: string[] = [];
+			for (const file of filesInThisDir) {
+				const fileWithPath = path + (path ? '/' : '') + file;
+				if (FS(`${basePath}/${fileWithPath}`).isDirSync()) {
+					if (file.startsWith('.')) continue;
+					allFiles = allFiles.concat(getFiles(basePath, fileWithPath));
+				} else {
+					allFiles.push(fileWithPath);
+				}
+			}
+			return allFiles;
+		};
+		let files = FS('server/chat-plugins').readdirSync();
+		// info always goes first so other plugins can shadow it, and private plugins are loaded separately
+		files = files.filter(file => file !== 'info.js' && file !== 'private');
 		files.unshift('info.js');
+		try {
+			if (FS('server/chat-plugins/private').isDirSync()) {
+				files = files.concat(getFiles('server/chat-plugins', 'private'));
+			}
+		} catch (err) {
+			if (err.code !== 'ENOENT') throw err;
+		}
 
 		for (const file of files) {
 			let plugin;
```

Comment:
Maybe the contents of this loop should be refactored out into a `loadPlugin` function, which could then be called inline.